File: scripts/process_meps_for_modeling.R
#!/usr/bin/env Rscript
# Download MEPS Stata FYC zips, drop competition exclusions, keep TOTEXPyy, write Parquet.
# Usage (from repo root):  Rscript scripts/process_meps_for_modeling.R [--download]

suppressPackageStartupMessages({
  if (!requireNamespace("haven", quietly = TRUE)) stop("install.packages('haven')")
  if (!requireNamespace("arrow", quietly = TRUE)) stop("install.packages('arrow')")
  if (!requireNamespace("dplyr", quietly = TRUE)) stop("install.packages('dplyr')")
  if (!requireNamespace("jsonlite", quietly = TRUE)) stop("install.packages('jsonlite')")
})

cmd <- commandArgs(trailingOnly = FALSE)
fn <- sub("^--file=", "", cmd[grep("^--file=", cmd)])
root <- normalizePath(file.path(dirname(fn), ".."), winslash = "/", mustWork = TRUE)
setwd(root)

source(file.path(root, "R", "meps_competition_exclusions.R"))

do_download <- "--download" %in% commandArgs(trailingOnly = TRUE)

year_map <- data.frame(
  calendar_year = c(2019, 2020, 2021, 2022, 2023),
  yy = c("19", "20", "21", "22", "23"),
  puf = c("h216", "h224", "h233", "h243", "h251"),
  stringsAsFactors = FALSE
)

base_url <- "https://meps.ahrq.gov/data_files/pufs/%s/%sdta.zip"

zip_dir <- file.path(root, "data", "raw", "stata_zips")
ex_dir <- file.path(root, "data", "raw", "stata")
out_dir <- file.path(root, "data", "processed")
dir.create(zip_dir, FALSE, TRUE)
dir.create(ex_dir, FALSE, TRUE)
dir.create(out_dir, FALSE, TRUE)

exc_all <- meps_expanded_exclusion_names()

rel_from_root <- function(path) {
  path <- normalizePath(path, winslash = "/", mustWork = TRUE)
  sub(paste0("^", root, "/"), "", path)
}

manifest <- vector("list", nrow(year_map))

for (i in seq_len(nrow(year_map))) {
  yr <- year_map$calendar_year[i]
  yy <- year_map$yy[i]
  puf <- year_map$puf[i]
  zipf <- file.path(zip_dir, paste0(puf, "dta.zip"))
  url <- sprintf(base_url, puf, puf)

  if (do_download || !file.exists(zipf)) {
    message("Downloading ", url)
    download.file(url, zipf, method = "libcurl", mode = "wb")
  }

  ex_sub <- file.path(ex_dir, puf)
  dir.create(ex_sub, FALSE, TRUE)
  unzip(zipf, exdir = ex_sub)
  dta <- list.files(ex_sub, pattern = "\\.dta$", ignore.case = TRUE, full.names = TRUE)
  if (!length(dta)) stop("No .dta in ", ex_sub)
  dta_path <- dta[[1]]

  message("Reading ", dta_path)
  df <- haven::read_dta(dta_path)
  df <- as.data.frame(df)

  target <- paste0("TOTEXP", yy)
  if (!target %in% names(df)) stop("Missing ", target)

  drop <- setdiff(intersect(names(df), exc_all), target)
  df <- df[, setdiff(names(df), drop), drop = FALSE]

  leaked <- meps_survey_design_present(names(df))
  if (length(leaked)) {
    stop("Survey design columns still present: ", paste(leaked, collapse = ", "))
  }

  outf <- file.path(out_dir, sprintf("meps_fyc_%s_for_modeling.parquet", yr))
  arrow::write_parquet(df, outf)

  manifest[[i]] <- list(
    calendar_year = as.character(yr),
    yy = yy,
    source_dta = rel_from_root(dta_path),
    out_parquet = rel_from_root(outf),
    n_rows = nrow(df),
    n_columns = ncol(df),
    n_columns_dropped = length(drop),
    target_column = target
  )
  message("Wrote ", outf, " (", nrow(df), " x ", ncol(df), ")")
}

jsonlite::write_json(
  manifest,
  file.path(out_dir, "processing_manifest.json"),
  auto_unbox = TRUE,
  pretty = TRUE
)
message("Wrote processing_manifest.json")

File: legacy/python/process_meps_for_modeling.py
#!/usr/bin/env python3
"""
Download MEPS full-year consolidated Stata files (HC-216 .. HC-251), drop competition-
excluded predictors, keep TOTEXPyy as the modeling target column, write Parquet.

Requires: pandas, pyarrow (see requirements.txt). Run from repo root with venv active:

  source .venv/bin/activate
  python scripts/process_meps_for_modeling.py --download

Raw ASCII .dat files are not used here (fixed-width layout); Stata PUFs are the reliable path.
"""

from __future__ import annotations

import argparse
import json
import subprocess
import sys
import zipfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from src.meps_competition_exclusions import (  # noqa: E402
    build_excluded_physical_names,
    survey_design_columns_present,
)

# HC PUF id -> two-digit MEPS variable year suffix on that file
YEAR_FILES: tuple[tuple[str, str, str], ...] = (
    ("2019", "19", "h216"),
    ("2020", "20", "h224"),
    ("2021", "21", "h233"),
    ("2022", "22", "h243"),
    ("2023", "23", "h251"),
)

BASE_URL = "https://meps.ahrq.gov/data_files/pufs/{puf}/{puf}dta.zip"


def download_zip(puf: str, dest_zip: Path) -> None:
    """Use curl so macOS / older Python SSL stores do not break HTTPS."""
    url = BASE_URL.format(puf=puf)
    dest_zip.parent.mkdir(parents=True, exist_ok=True)
    print(f"Downloading {url} -> {dest_zip}")
    subprocess.run(
        ["curl", "-fsSL", "-o", str(dest_zip), url],
        check=True,
        timeout=600,
    )


def extract_dta(zip_path: Path, extract_root: Path, puf: str) -> Path:
    dest_parent = extract_root / puf
    dest_parent.mkdir(parents=True, exist_ok=True)
    with zipfile.ZipFile(zip_path, "r") as zf:
        dta_members = [n for n in zf.namelist() if n.lower().endswith(".dta")]
        if len(dta_members) != 1:
            raise RuntimeError(f"{zip_path}: expected exactly one .dta, got {dta_members!r}")
        member = dta_members[0]
        zf.extract(member, path=dest_parent)
        return dest_parent / member


def process_one(
    calendar_year: str,
    yy: str,
    dta_path: Path,
    out_dir: Path,
    excluded: frozenset[str],
) -> dict:
    import pandas as pd

    target = f"TOTEXP{yy}"
    print(f"Reading {dta_path} ({calendar_year}, target={target})")
    df = pd.read_stata(dta_path, convert_categoricals=False)

    if target not in df.columns:
        raise KeyError(f"Missing target column {target!r} in {dta_path}")

    drop = [c for c in df.columns if c in excluded and c != target]
    kept = [c for c in df.columns if c not in drop]
    df_out = df[kept].copy()

    leaked = survey_design_columns_present(df_out.columns)
    if leaked:
        raise RuntimeError(f"Survey design columns still present (should be excluded): {leaked[:20]}")

    out_path = out_dir / f"meps_fyc_{calendar_year}_for_modeling.parquet"
    out_dir.mkdir(parents=True, exist_ok=True)
    df_out.to_parquet(out_path, index=False, engine="pyarrow")
    print(f"Wrote {out_path} shape={df_out.shape} dropped={len(drop)}")

    def rel(p: Path) -> str:
        try:
            return str(p.resolve().relative_to(ROOT.resolve()))
        except ValueError:
            return str(p)

    return {
        "calendar_year": calendar_year,
        "yy": yy,
        "source_dta": rel(dta_path),
        "out_parquet": rel(out_path),
        "n_rows": int(len(df_out)),
        "n_columns": int(df_out.shape[1]),
        "n_columns_dropped": len(drop),
        "target_column": target,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--download",
        action="store_true",
        help="Fetch official Stata zips from MEPS (needed unless .dta already extracted).",
    )
    parser.add_argument(
        "--zip-dir",
        type=Path,
        default=ROOT / "data" / "raw" / "stata_zips",
        help="Where to store downloaded zips.",
    )
    parser.add_argument(
        "--extract-dir",
        type=Path,
        default=ROOT / "data" / "raw" / "stata",
        help="Where to extract .dta files.",
    )
    parser.add_argument(
        "--out-dir",
        type=Path,
        default=ROOT / "data" / "processed",
        help="Where to write Parquet outputs.",
    )
    args = parser.parse_args()

    excluded = build_excluded_physical_names()
    manifest: list[dict] = []

    for calendar_year, yy, puf in YEAR_FILES:
        zip_path = args.zip_dir / f"{puf}dta.zip"
        if args.download or not zip_path.is_file():
            download_zip(puf, zip_path)
        dta_path = extract_dta(zip_path, args.extract_dir, puf)
        info = process_one(calendar_year, yy, dta_path, args.out_dir, excluded)
        manifest.append(info)

    manifest_path = args.out_dir / "processing_manifest.json"
    manifest_path.write_text(json.dumps(manifest, indent=2), encoding="utf-8")
    print(f"Wrote {manifest_path}")


if __name__ == "__main__":
    main()
